core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -545,9 +551,18 @@ class LogManager(logDirs: Array[File],
         if (!isNew && offlineLogDirs.nonEmpty)
           throw new KafkaStorageException(s"Can not create log for $topicPartition because log directories ${offlineLogDirs.mkString(",")} are offline")
 
-        val dataDir = nextLogDir()
+        val logDir = {
+          val preferredLogDir = preferredLogDirs.get(topicPartition)
+          if (preferredLogDir != null)
+            preferredLogDir
+          else
+            nextLogDir().getAbsolutePath
+        }
+        if (!isLogDirOnline(logDir))
+          throw new KafkaStorageException(s"Can not create log for $topicPartition because log directory $logDir is offline")
```

Comment:
With this, it's possible for a created log not to be in logs. Then, we won't find this log in LogManager.getLog(). When serving requests like OffsetByTimestamp, we will probably return UnknownTopicPartitionException, but ideally we want to return KafkaStorageException.